lib/Sema/TypeCheckAttr.cpp

```diff
@@ -5513,3 +5514,51 @@ void TypeChecker::checkClosureAttributes(ClosureExpr *closure) {
     checker.visit(attr);
   }
 }
+
+void AttributeChecker::visitHasAsyncAlternativeAttr(
+    HasAsyncAlternativeAttr *attr) {
+  // TODO: We should lookup and store the referenced function decl now that they
+  //       have all been parsed. From there, we can verify that the two function
+  //       types match.
+
+  // We have two function decls to worry about here for now.
+  // There will be a third for the found async function decl.
+  // We have the decl that the attribute is attached to
+  //  callingFunctionDecl
+  // We have the completion handler decl passed into the function
+  //  completionHandlerDecl
+  auto callingFunctionDecl = cast<AbstractFunctionDecl>(D);
+  if (callingFunctionDecl->hasAsync()) {
+    diagnose(callingFunctionDecl->getAsyncLoc(), diag::pound_error,
+             "'@hasAsyncAlternative' attached to async function");
+    return;
+  }
+
+  const ParameterList *params = callingFunctionDecl->getParameters();
+  if (!params) // TODO: Figure out how to exercise this and emit an error
+    llvm_unreachable("The parameter list is a nullptr");
```

Comment:
From what I can tell, it's always set. All callers that I saw assume that it is as well.

Would be nice to clean up the AST construction at some point (ie. it should be obvious that this is true, but it really isn't since setParameters is public). In general there seems to be a fair bit of assumed postconditions that should be enforced better.